libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,363 @@
+{
+	return std::visit(util::GenericVisitor{
+		[](FunctionCallReturnLabelSlot const& _ret) -> std::string { return "RET[" + _ret.call.get().functionName.name.str() + "]"; },
+		[](FunctionReturnLabelSlot const&) -> std::string { return "RET"; },
+		[](VariableSlot const& _var) { return _var.variable.get().name.str(); },
+		[](LiteralSlot const& _lit) { return util::toCompactHexWithPrefix(_lit.value); },
+		[](TemporarySlot const& _tmp) -> std::string { return "TMP[" + _tmp.call.get().functionName.name.str() + ", " + std::to_string(_tmp.index) + "]"; },
+		[](JunkSlot const&) -> std::string { return "JUNK"; }
+	}, _slot);
+}
+
+inline std::string stackToString(Stack const& _stack)
+{
+	std::string result("[ ");
+	for (auto const& slot: _stack)
+		result += stackSlotToString(slot) + ' ';
+	result += ']';
+	return result;
+}
+
+
+// Abstraction of stack shuffling operations. Can be defined as actual concept once we switch to C++20.
+/*
+template<typename ShuffleOperations>
+concept ShuffleOperationConcept = requires(ShuffleOperations ops, size_t sourceOffset, size_t targetOffset, size_t depth) {
+	// Returns true, iff the current slot at sourceOffset in source layout is a suitable slot at targetOffset.
+	{ ops.isCompatible(sourceOffset, targetOffset) } -> std::convertible_to<bool>;
+	// Returns true, iff the slots at the two given source offsets are identical.
+	{ ops.sourceIsSame(sourceOffset, sourceOffset) } -> std::convertible_to<bool>;
+	// Returns a positive integer n, if the slot at the given source offset needs n more copies.
+	// Returns a negative integer -n, if the slot at the given source offsets occurs n times too many.
+	// Returns zero if the amount of occurrences, in the current source layout, of the slot at the given source offset matches the desired amount of occurrences in the target.
+	{ ops.sourceMultiplicity(sourceOffset) } -> std::convertible_to<int>;
+	// Returns a positive integer n, if the slot at the given target offset needs n more copies.
+	// Returns a negative integer -n, if the slot at the given target offsets occurs n times too many.
+	// Returns zero if the amount of occurrences, in the current source layout, of the slot at the given target offset matches the desired amount of occurrences in the target.
+	{ ops.targetMultiplicity(targetOffset) } -> std::convertible_to<int>;
+	// Returns true, iff any slot is compatible with the given target offset.
+	{ ops.targetIsArbitrary(targetOffset) } -> std::convertible_to<bool>;
+	// Returns the number of slots in the source layout.
+	{ ops.sourceSize() } -> std::convertible_to<size_t>;
+	// Returns the number of slots in the target layout.
+	{ ops.targetSize() } -> std::convertible_to<size_t>;
+	// Swaps the top most slot in the source with the slot at depth.
+	{ ops.swap(depth) };
+	// Pops the top most slot in the source.
+	{ ops.pop() };
+	// Dups or pushes the slot that is supposed to end up at the given target offset.
```

Comment:
So it takes the slot from the target and pushes it to the source / current?

Reply:
Yes. Or in practice, rather it looks for the least deep occurrence of the desired target slot in the source and duplicates that or, if that's not possible and the slot can be freely generated, pushes a new copy of it.